### PR TITLE
Add reconnect logic

### PR DESCRIPTION
**Pull request by jiancheung**
https://clever.atlassian.net/browse/DISTRICTS-261
Pretty much followed the instructions at https://github.com/mikespook/gearman-go/issues/6 and https://github.com/draxil/gearman-go-stubborn/blob/master/worker/stubborn/stubborn.go

Verified in staging that this works.

### CODE REVIEW

**Comment by jiancheung** [comment]
cc @dfuentes

File: baseworker/worker.go [hunk 1]
```diff
@@ -75,18 +92,21 @@ func NewWorker(name string, fn JobFunc) *Worker {
...
-			if !opErr.Temporary() {
-				proc, err := os.FindProcess(os.Getpid())
-				if err != nil {
-					lg.InfoD("err-getpid", logger.M{"error": err.Error()})
-				}
-				if err := proc.Signal(os.Interrupt); err != nil {
-					lg.InfoD("err-interrupt", logger.M{"error": err.Error()})
-				}
+		// Try to reconnect if it is a disconnect error
+		wdc, wdcok := e.(*gearmanWorker.WorkerDisconnectError)
```
Comment (natebrennand) [inline baseworker/worker.go]: [nit] can this just be `ok`, or `isWdc`?

File: baseworker/worker.go [hunk 2]
```diff
@@ -75,18 +92,21 @@ func NewWorker(name string, fn JobFunc) *Worker {
...
+		if wdcok {
+			lg.InfoD("err-disconnected-and-reconnecting", logger.M{"name": name, "error": e.Error()})
+			r := retrier.New(retrier.ExponentialBackoff(5, 200*time.Millisecond), nil)
+			if rc_err := r.Run(wdc.Reconnect); rc_err != nil {
+				lg.InfoD("err-disconnected-fully", logger.M{"name": name, "error": rc_err.Error()})
+				defaultErrorHandler(rc_err)
+				return
 			}
+			lg.InfoD("reconnected", logger.M{"name": name})
```
Comment (natebrennand) [inline baseworker/worker.go]: s/`reconnected`/`gearman-reconnected` ?

File: baseworker/worker.go [hunk 3]
```diff
@@ -75,18 +92,21 @@ func NewWorker(name string, fn JobFunc) *Worker {
...
-				if err := proc.Signal(os.Interrupt); err != nil {
-					lg.InfoD("err-interrupt", logger.M{"error": err.Error()})
-				}
+		// Try to reconnect if it is a disconnect error
+		wdc, wdcok := e.(*gearmanWorker.WorkerDisconnectError)
+		if wdcok {
+			lg.InfoD("err-disconnected-and-reconnecting", logger.M{"name": name, "error": e.Error()})
+			r := retrier.New(retrier.ExponentialBackoff(5, 200*time.Millisecond), nil)
+			if rc_err := r.Run(wdc.Reconnect); rc_err != nil {
+				lg.InfoD("err-disconnected-fully", logger.M{"name": name, "error": rc_err.Error()})
```
Comment (natebrennand) [inline baseworker/worker.go]: `lg.ErrorD` or `lg.CriticalD` ?

File: baseworker/worker.go [hunk 4]
```diff
@@ -67,6 +69,21 @@ func defaultSigtermHandler(worker *Worker) {
...
+func defaultErrorHandler(e error) {
+	lg.InfoD("gearman-error", logger.M{"error": e.Error()})
+	if opErr, ok := e.(*net.OpError); ok {
+		if !opErr.Temporary() {
+			proc, err := os.FindProcess(os.Getpid())
+			if err != nil {
+				lg.InfoD("err-getpid", logger.M{"error": err.Error()})
+			}
+			if err := proc.Signal(os.Interrupt); err != nil {
+				lg.InfoD("err-interrupt", logger.M{"error": err.Error()})
```
Comment (natebrennand) [inline baseworker/worker.go]: `lg.CriticalD`?

File: baseworker/worker.go [hunk 5]
```diff
@@ -67,6 +69,21 @@ func defaultSigtermHandler(worker *Worker) {
 	os.Exit(0)
 }
 
+func defaultErrorHandler(e error) {
+	lg.InfoD("gearman-error", logger.M{"error": e.Error()})
+	if opErr, ok := e.(*net.OpError); ok {
+		if !opErr.Temporary() {
+			proc, err := os.FindProcess(os.Getpid())
+			if err != nil {
+				lg.InfoD("err-getpid", logger.M{"error": err.Error()})
```
Comment (natebrennand) [inline baseworker/worker.go]: probs `lg.CriticalD` too?

**Comment by natebrennand** [comment]
just some logging & naming nits, otherwise LGTM

**Pull request merged**